### PR TITLE
https://github.com/DEX-Company/starfish-java/issues/113

### PR DESCRIPTION
**Pull request by AyushVerma2**
Ensure getContentHash uses getContentStream

### CODE REVIEW

File: src/test/java/sg/dex/starfish/impl/TestResources.java [hunk 1]
```diff
@@ -15,18 +17,12 @@
...
         byte[] bs = dataAsset.getContent();
-        String s = new String(bs, StandardCharsets.UTF_8);
-        Assertions.assertEquals("Hello Starfish", s);
-
-        Assertions.assertNotNull(dataAsset.getMetadataString());
-        Assertions.assertEquals(14, dataAsset.getContentSize());
-
         assertFalse(dataAsset.getMetadata().containsKey(Constant.CONTENT_HASH));
         dataAsset = dataAsset.includeContentHash();
-        Assertions.assertEquals("e4be09f07f5665ecacc078223e86c1dba18b38a3e07a3d575167b5ba7a1821d1", dataAsset.getMetadata().get(Constant.CONTENT_HASH));
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/TestResources.java:29]: Please test for the the correct hash value!

File: src/test/java/sg/dex/starfish/impl/TestResources.java [hunk 2]
```diff
@@ -15,18 +17,12 @@
 public class TestResources {
 
     @Test
-    public void testResourceAssetWithContentHash() {
+    public void testResourceAssetWithContentHash() throws DigestException {
         DataAsset dataAsset = ResourceAsset.create("assets/hello.txt");
         byte[] bs = dataAsset.getContent();
-        String s = new String(bs, StandardCharsets.UTF_8);
-        Assertions.assertEquals("Hello Starfish", s);
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/TestResources.java:22]: Why are we deleting these lines of tests?

File: src/main/java/sg/dex/starfish/DataAsset.java [hunk 3]
```diff
@@ -97,7 +103,7 @@ default String getContentHash() {
      * @return respective data asset sub class.
      * @throws UnsupportedOperationException if the Asset type is not DataAsset
      */
-    default DataAsset includeContentHash() {
+    default DataAsset includeContentHash() throws DigestException {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/DataAsset.java]: Should this really be throwing a DigestException? Under what circumstances can this happen? Perhaps we should be catching and re-throwing if this happens?

File: src/main/java/sg/dex/crypto/Hash.java [hunk 4]
```diff
@@ -123,4 +127,34 @@ public static String sha3_256String(String string) {
...
+     * A MessageDigest object starts out initialized.
+     * The data is processed through it using the update methods.
+     * At any point reset can be called to reset the digest.
+     * Once all the data to be updated has been updated, one of the digest
+     * methods should be called to complete the hash computation.
+     *
+     * @param inputStream
+     * @return
+     */
+    public static String checkSum(InputStream inputStream) throws DigestException {
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java]: Can we use a better name? It isn't really a checksum.

File: src/main/java/sg/dex/crypto/Hash.java [hunk 5]
```diff
@@ -123,4 +127,34 @@ public static String sha3_256String(String string) {
...
+
+            SHA3.DigestSHA3 md = new SHA3.Digest256();
+            //Using MessageDigest update() method to provide input
+            byte[] buffer = new byte[8192];
+            int numOfBytesRead;
+            while ((numOfBytesRead = inputStream.read(buffer)) > 0) {
+                md.update(buffer, 0, numOfBytesRead);
+            }
+            byte[] hash = md.digest();
+            checksum = new BigInteger(1, hash).toString(16); //don't use this, truncates leading zero
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java]: Don't use BigInteger here. Use a static utility method to convert byte[] to hex String, I think one exists already.

File: src/main/java/sg/dex/crypto/Hash.java [hunk 6]
```diff
@@ -123,4 +127,34 @@ public static String sha3_256String(String string) {
         return Hex.toString(sha3_256(string));
     }
 
+    /**
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java:131]: This doesn't make sense as a comment for this function. Document what a function does, not the implementation details.

File: src/integration-test/java/developerTC/TestAssetRegistration_08.java [hunk 7]
```diff
@@ -124,10 +126,9 @@ public void testHashForRemoteAsset() throws IOException {
         Asset memoryAsset = MemoryAsset.create(content.getBytes());
 
         DataAsset dataAsset = remoteAgent.uploadAsset(memoryAsset);
-        DataAsset dataAssetWithHash = dataAsset.includeContentHash();
+        dataAsset.includeContentHash();
 
-        Assertions.assertEquals(Hex.toString(Hash.sha3_256(dataAsset.getContent())), Hex.toString(Hash.sha3_256(content)));
-//        assertEquals(Hex.toString(Hash.sha3_256(dataAssetWithHash.getMetadata().get(CONTENT_HASH).toString())), Hex.toString(Hash.sha3_256(content)));
+        Assertions.assertEquals(Hash.checkSum(dataAsset.getContentStream()), dataAsset.getMetadata().get(Constant.CONTENT_HASH));
```
Comment (mikera) [inline src/integration-test/java/developerTC/TestAssetRegistration_08.java]: How do we know this worked? Use the exact string hash value to check for correctness.

File: src/main/java/sg/dex/crypto/Hash.java [hunk 8]
```diff
@@ -123,4 +127,34 @@ public static String sha3_256String(String string) {
...
+            //Using MessageDigest update() method to provide input
+            byte[] buffer = new byte[8192];
+            int numOfBytesRead;
+            while ((numOfBytesRead = inputStream.read(buffer)) > 0) {
+                md.update(buffer, 0, numOfBytesRead);
+            }
+            byte[] hash = md.digest();
+            checksum = new BigInteger(1, hash).toString(16); //don't use this, truncates leading zero
+        } catch (IOException e) {
+            throw new DigestException("couldn't make digest of partial content");
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java]: What does "partial content" mean? I think DigestException is the wrong exception type here, isn't this an IO problem?

File: src/main/java/sg/dex/crypto/Hash.java [hunk 9]
```diff
@@ -4,7 +4,11 @@
 import org.bouncycastle.jcajce.provider.digest.SHA3;
 import sg.dex.starfish.util.Hex;
 
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java:10]: Don't use BigInteger unless absolutely necessary. Slow and error prone due to sign issues etc.

File: src/main/java/sg/dex/crypto/Hash.java [hunk 10]
```diff
@@ -123,4 +127,34 @@ public static String sha3_256String(String string) {
...
+            while ((numOfBytesRead = inputStream.read(buffer)) > 0) {
+                md.update(buffer, 0, numOfBytesRead);
+            }
+            byte[] hash = md.digest();
+            checksum = new BigInteger(1, hash).toString(16); //don't use this, truncates leading zero
+        } catch (IOException e) {
+            throw new DigestException("couldn't make digest of partial content");
+        }
+        assert checksum != null;
+        return checksum.trim();
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java]: How can the checksum be null? How can it need trimming? This look like issues that shouldn't be occurring.

**Review by mikera — CHANGES_REQUESTED** [review]

File: src/test/java/sg/dex/starfish/impl/TestResources.java (hunk 1 above)
Comment (AyushVerma2) [reply src/test/java/sg/dex/starfish/impl/TestResources.java:29]: Fixed

File: src/test/java/sg/dex/starfish/impl/TestResources.java (hunk 2 above)
Comment (AyushVerma2) [reply src/test/java/sg/dex/starfish/impl/TestResources.java:22]: updated

File: src/main/java/sg/dex/crypto/Hash.java (hunk 4 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/crypto/Hash.java]: update name as computeHashWithSHA3

File: src/main/java/sg/dex/crypto/Hash.java (hunk 5 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/crypto/Hash.java]: done.

File: src/main/java/sg/dex/crypto/Hash.java (hunk 8 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/crypto/Hash.java]: yes, updated as IO exception

File: src/main/java/sg/dex/crypto/Hash.java (hunk 10 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/crypto/Hash.java]: done

File: src/test/java/sg/dex/starfish/impl/TestResources.java (hunk 1 above)
Comment (mikera) [reply src/test/java/sg/dex/starfish/impl/TestResources.java:29]: Should test for a specific hex string - why are we removing this line?

File: src/main/java/sg/dex/crypto/Hash.java [hunk 11]
```diff
@@ -123,4 +128,36 @@ public static String sha3_256String(String string) {
...
+            //Using MessageDigest update() method to provide input
+            byte[] buffer = new byte[8192];
+            int numOfBytesRead;
+            while ((numOfBytesRead = inputStream.read(buffer)) > 0) {
+                md.update(buffer, 0, numOfBytesRead);
+            }
+            byte[] hash = md.digest();
+            checksum =  Hex.toString(hash);
+        } catch (IOException e) {
+            throw new IOException("Couldn't compute the hash  of partial content");
```
Comment (mikera) [inline src/main/java/sg/dex/crypto/Hash.java]: What is "partial content"? Error message doesn't make sense. Just remove the whole try / catch block

File: src/test/java/sg/dex/starfish/impl/TestResources.java (hunk 1 above)
Comment (AyushVerma2) [reply src/test/java/sg/dex/starfish/impl/TestResources.java:29]: this value was not clear to user , so I calculated the hex value using SHA3 and used the same for validation.Variable expected is the same value as e4be09f07f5665ecacc078223e86c1dba18b38a3e07a3d575167b5ba7a1821d1.

**Comment by AyushVerma2** [comment]
done.

**Pull request closed without merging**